### ISSUE TITLE
EF Core 2.0 - InMemoryDatabase Testing - Changing a value of an attribute is Updating the database record

### ISSUE THREAD

**Issue by JRBonnema** [dotnet/efcore#10606]
I observed what I consider unexpected behaviour when doing integration testing using InMemoryDatabase testing. It seems that line 4 in the TestMethod below is actually updating the database record from OPEN to CLOSED. It was intended to be only a preparation for the actual update using the repository that for simplicity I have left out here. I'm aware that quote "InMemory is designed to be a general purpose database for testing, and is not designed to mimic a relational database", but in my opinion this behaviour is not what I expect from a general purpose database. I would like to hear from you if you consider this a bug. If not how can I test my update procedures using InMemoryDatabase ?

```C#
        [TestMethod]
        public void Authors_ChangeStatusCode_StrangeBehaviour()
        {
            Author authToBeUpdated = _repository.getAuthor("Postma");
            Guid internalCode = authToBeUpdated.InternalCode;
            Assert.IsTrue(authToBeUpdated.StatusCode == "OPEN");

        4)  authToBeUpdated.StatusCode = "CLOSED";

            Author authFound1 = _repository.get(authToBeUpdated.InternalCode);
            Assert.IsTrue(authFound1.StatusCode == "CLOSED");
            Author authFound2 = _repository.getAuthor(authToBeUpdated.Name);
            Assert.IsTrue(authFound2.StatusCode == "CLOSED");
        }
```
        Part of creating the record - works correctly.
```C#
        private Author createAuthorRecord(string authorName, string externalCode)
        {
            Author auth = new Author();
            auth.InternalCode = Guid.NewGuid();
            auth.ExternalCode = externalCode;
            auth.StatusCode = "OPEN";
            auth.Name = authorName;

           ....

            return auth;
        }
```
      Base class for constructing the  .UseInMemoryDatabase

```C#
      public abstract class AbstractTestBase
      {
        protected readonly BooksDbContext _testDbContext;

        public AbstractTestBase()
        {
            DbContextOptions<BooksDbContext> _options
                = new DbContextOptionsBuilder<BooksDbContext>()
                        .UseInMemoryDatabase(databaseName: Guid.NewGuid().ToString())
                        .Options;

            _testDbContext = new BooksDbContext(_options);
            _testDbContext.Database.EnsureCreated();
        }

        public virtual void Dispose()
        {
            _testDbContext.Database.EnsureDeleted();
            _testDbContext.Dispose();
        }
```

### Further technical details
EF Core version: 2.0.1
Database Provider: InMemoryDatabase
Operating system: Windows 10
IDE: (e.g. Visual Studio 2017 15.5.2)

**Comment by ajcvickers** [comment]
@JRBonnema Without seeing the repository code, I would guess that that the get method is returning the same instance, probably because the context is tracking it. When EF does executes a query that returns the an entity that is already tracked, then the tracked instance is returned, which in this case already contains the modification made.

**Comment by JRBonnema** [comment]
@ajcvickers  You diagnosis was entiry correct. I found out that In [TestInitialize] I added three records without using the repository, that were not detached after adding the records to the InMemoryDatabase, resulting in the behaviour you describe above. 

When I added this three records using the repository.add() function the problem went away.  Sorry for the inconvenience and let me use this occasion to complement you and your team colleagues on the great work your doing with Entity Framework.
